File: apps/golang/internal/graphql/domain/bkm/bkm_report.go
package bkm

// ============================================================================
// BKM Potong Buah Report — Domain types used by the report service
// These mirror the GraphQL schema types in bkm_report.graphqls
// ============================================================================

// BkmPotongBuahFilter is the filter input for the report query.
type BkmPotongBuahFilter struct {
	Periode   int32   `json:"periode"`
	CompanyID *string `json:"companyId,omitempty"`
	Estate    *string `json:"estate,omitempty"`
	Divisi    *string `json:"divisi,omitempty"`
	Blok      *string `json:"blok,omitempty"`
}

// BkmPotongBuahDetail represents a single detail row in the report.
type BkmPotongBuahDetail struct {
	Tanggal *string  `json:"tanggal,omitempty"`
	Mandor  *string  `json:"mandor,omitempty"`
	NIK     *string  `json:"nik,omitempty"`
	Nama    *string  `json:"nama,omitempty"`
	Qty     *float64 `json:"qty,omitempty"`
	Satuan  *string  `json:"satuan,omitempty"`
	Jumlah  *float64 `json:"jumlah,omitempty"`
}

// BkmPotongBuahBlock represents block-level aggregation.
type BkmPotongBuahBlock struct {
	Blok         string                 `json:"blok"`
	TotalQty     float64                `json:"totalQty"`
	TotalJumlah  float64                `json:"totalJumlah"`
	TotalRecords int32                  `json:"totalRecords"`
	Details      []*BkmPotongBuahDetail `json:"details"`
}

// BkmPotongBuahDivision represents division-level aggregation.
type BkmPotongBuahDivision struct {
	Divisi       string                `json:"divisi"`
	TotalQty     float64               `json:"totalQty"`
	TotalJumlah  float64               `json:"totalJumlah"`
	TotalRecords int32                 `json:"totalRecords"`
	Blocks       []*BkmPotongBuahBlock `json:"blocks"`
}

// BkmPotongBuahEstate represents estate-level aggregation.
type BkmPotongBuahEstate struct {
	Periode      int32                    `json:"periode"`
	Estate       string                   `json:"estate"`
	TotalRecords int32                    `json:"totalRecords"`
	TotalQty     float64                  `json:"totalQty"`
	TotalJumlah  float64                  `json:"totalJumlah"`
	Divisions    []*BkmPotongBuahDivision `json:"divisions"`
}

// BkmPotongBuahSummary is the top-level report result.
type BkmPotongBuahSummary struct {
	TotalRecords int32                  `json:"totalRecords"`
	TotalQty     float64                `json:"totalQty"`
	TotalJumlah  float64                `json:"totalJumlah"`
	Estates      []*BkmPotongBuahEstate `json:"estates"`
}

// rawPotongBuahRow is the flat row returned from the SQL query.
type RawPotongBuahRow struct {
	Periode int32
	Estate  string
	Divisi  string
	Blok    string
	Tanggal *string
	Mandor  *string
	NIK     *string
	Nama    *string
	Qty     float64
	Satuan  *string
	Jumlah  float64
}

// ============================================================================
// Flat Report Models
// ============================================================================

type BkmPotongBuahFlatItem struct {
	Tanggal     string   `json:"tanggal"`
	CompanyCode string   `json:"companyCode"`
	CompanyName string   `json:"companyName"`
	Estate      string   `json:"estate"`
	Divisi      string   `json:"divisi"`
	Blok        string   `json:"blok"`
	NIK         *string  `json:"nik"`
	Nama        *string  `json:"nama"`
	QtyP1       *float64 `json:"qtyp1"`
	SatP1       *string  `json:"satp1"`
	QtyP2       *float64 `json:"qtyp2"`
	SatP2       *string  `json:"satp2"`
	Qty         *float64 `json:"qty"`
	Satuan      *string  `json:"satuan"`
	Jumlah      *float64 `json:"jumlah"`
}

type BkmPotongBuahKPI struct {
	TotalQty    float64 `json:"totalQty"`
	TotalJumlah float64 `json:"totalJumlah"`
	TotalHk     int32   `json:"totalHk"`
	Bgm         float64 `json:"bgm"`
	OutputPerHk float64 `json:"outputPerHk"`
}

type BkmPotongBuahFlatResponse struct {
	Data    []*BkmPotongBuahFlatItem `json:"data"`
	Total   int32                    `json:"total"`
	Summary *BkmPotongBuahKPI        `json:"summary"`
}

type BkmPotongBuahDailyPoint struct {
	Date        string  `json:"date"`
	OutputQty   float64 `json:"outputQty"`
	TotalJumlah float64 `json:"totalJumlah"`
	WorkerCount int32   `json:"workerCount"`
}

type BkmPotongBuahOutputPoint struct {
	Name      string  `json:"name"`
	OutputQty float64 `json:"outputQty"`
}

type BkmPotongBuahHarvesterPoint struct {
	NIK       string  `json:"nik"`
	Name      string  `json:"name"`
	OutputQty float64 `json:"outputQty"`
}

type BkmPotongBuahAnalytics struct {
	TotalRecords int32                          `json:"totalRecords"`
	Summary      *BkmPotongBuahKPI              `json:"summary"`
	Daily        []*BkmPotongBuahDailyPoint     `json:"daily"`
	Companies    []*BkmPotongBuahOutputPoint    `json:"companies"`
	Estates      []*BkmPotongBuahOutputPoint    `json:"estates"`
	Divisions    []*BkmPotongBuahOutputPoint    `json:"divisions"`
	Blocks       []*BkmPotongBuahOutputPoint    `json:"blocks"`
	Harvesters   []*BkmPotongBuahHarvesterPoint `json:"harvesters"`
}

File: apps/web/lib/apollo/queries/bkm-report.ts
import { gql } from '@apollo/client';

export const GET_BKM_POTONG_BUAH_REPORT = gql`
  query BkmPotongBuahReport($filter: BkmPotongBuahFilter!) {
    bkmPotongBuahReport(filter: $filter) {
      totalRecords
      totalQty
      totalJumlah
      estates {
        periode
        estate
        totalRecords
        totalQty
        totalJumlah
        divisions {
          divisi
          totalQty
          totalJumlah
          totalRecords
          blocks {
            blok
            totalQty
            totalJumlah
            totalRecords
            details {
              tanggal
              mandor
              nik
              nama
              qty
              satuan
              jumlah
            }
          }
        }
      }
    }
  }
`;

export interface BkmPotongBuahDetail {
  tanggal: string;
  mandor: string;
  nik: string;
  nama: string;
  qty: number;
  satuan: string;
  jumlah: number;
}

export interface BkmPotongBuahBlock {
  blok: string;
  totalQty: number;
  totalJumlah: number;
  totalRecords: number;
  details: BkmPotongBuahDetail[];
}

export interface BkmPotongBuahDivision {
  divisi: string;
  totalQty: number;
  totalJumlah: number;
  totalRecords: number;
  blocks: BkmPotongBuahBlock[];
}

export interface BkmPotongBuahEstate {
  periode: number;
  estate: string;
  totalRecords: number;
  totalQty: number;
  totalJumlah: number;
  divisions: BkmPotongBuahDivision[];
}

export interface BkmPotongBuahSummary {
  totalRecords: number;
  totalQty: number;
  totalJumlah: number;
  estates: BkmPotongBuahEstate[];
}

export interface BkmPotongBuahFilter {
  periode: number;
  companyId?: string;
  estate?: string;
  divisi?: string;
  blok?: string;
}

export interface BkmPotongBuahReportData {
  bkmPotongBuahReport: BkmPotongBuahSummary;
}

export interface BkmPotongBuahReportVars {
  filter: BkmPotongBuahFilter;
}

export const GET_BKM_POTONG_BUAH_FLAT = gql`
  query BkmPotongBuahFlat($filter: BkmPotongBuahFilter!, $page: Int!, $limit: Int!) {
    bkmPotongBuahFlat(filter: $filter, page: $page, limit: $limit) {
      data {
        tanggal
        companyCode
        companyName
        estate
        divisi
        blok
        nik
        nama
        qtyp1
        satp1
        qtyp2
        satp2
        qty
        satuan
        jumlah
      }
      total
      summary {
        totalQty
        totalJumlah
        totalHk
        bgm
        outputPerHk
      }
    }
  }
`;

export const GET_BKM_POTONG_BUAH_FLAT_SUMMARY = gql`
  query BkmPotongBuahFlatSummary($filter: BkmPotongBuahFilter!, $page: Int!, $limit: Int!) {
    bkmPotongBuahFlat(filter: $filter, page: $page, limit: $limit) {
      total
      summary {
        totalQty
        totalJumlah
        totalHk
        bgm
        outputPerHk
      }
    }
  }
`;

export const GET_BKM_POTONG_BUAH_ANALYTICS = gql`
  query BkmPotongBuahAnalytics($filter: BkmPotongBuahFilter!, $topN: Int) {
    bkmPotongBuahAnalytics(filter: $filter, topN: $topN) {
      totalRecords
      summary {
        totalQty
        totalJumlah
        totalHk
        bgm
        outputPerHk
      }
      daily {
        date
        outputQty
        totalJumlah
        workerCount
      }
      companies {
        name
        outputQty
      }
      estates {
        name
        outputQty
      }
      divisions {
        name
        outputQty
      }
      blocks {
        name
        outputQty
      }
      harvesters {
        nik
        name
        outputQty
      }
    }
  }
`;

export interface BkmPotongBuahFlatItem {
  tanggal: string;
  companyCode: string;
  companyName: string;
  estate: string;
  divisi: string;
  blok: string;
  nik: string;
  nama: string;
  qtyp1: number;
  satp1: string;
  qtyp2: number;
  satp2: string;
  qty: number;
  satuan: string;
  jumlah: number;
}

export interface BkmPotongBuahKPI {
  totalQty: number;
  totalJumlah: number;
  totalHk: number;
  bgm: number;
  outputPerHk: number;
}

export interface BkmPotongBuahFlatResponse {
  data: BkmPotongBuahFlatItem[];
  total: number;
  summary: BkmPotongBuahKPI;
}

export interface BkmPotongBuahFlatSummaryResponse {
  total: number;
  summary: BkmPotongBuahKPI;
}

export interface BkmPotongBuahDailyPoint {
  date: string;
  outputQty: number;
  totalJumlah: number;
  workerCount: number;
}

export interface BkmPotongBuahOutputPoint {
  name: string;
  outputQty: number;
}

export interface BkmPotongBuahHarvesterPoint {
  nik: string;
  name: string;
  outputQty: number;
}

export interface BkmPotongBuahAnalyticsResponse {
  totalRecords: number;
  summary: BkmPotongBuahKPI;
  daily: BkmPotongBuahDailyPoint[];
  companies: BkmPotongBuahOutputPoint[];
  estates: BkmPotongBuahOutputPoint[];
  divisions: BkmPotongBuahOutputPoint[];
  blocks: BkmPotongBuahOutputPoint[];
  harvesters: BkmPotongBuahHarvesterPoint[];
}

export interface BkmPotongBuahFlatData {
  bkmPotongBuahFlat: BkmPotongBuahFlatResponse;
}

export interface BkmPotongBuahFlatSummaryData {
  bkmPotongBuahFlat: BkmPotongBuahFlatSummaryResponse;
}

export interface BkmPotongBuahAnalyticsData {
  bkmPotongBuahAnalytics: BkmPotongBuahAnalyticsResponse;
}

export interface BkmPotongBuahFlatVars {
  filter: BkmPotongBuahFilter;
  page: number;
  limit: number;
}

export interface BkmPotongBuahAnalyticsVars {
  filter: BkmPotongBuahFilter;
  topN?: number;
}
